src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -251,128 +244,145 @@ private static string ComputeHash(string data, string algorithm)
+            internal const int NonceCount = 1;
 
-            public DigestResponse(string challenge)
+            internal DigestResponse(string challenge)
             {
-                Nonce = null;
-                NonceCount = 1;
-
                 Parse(challenge);
             }
 
-            private unsafe string GetNextKey(ref char* p)
+            private string GetNextKey(string data, int currentIndex, out int parsedIndex)
             {
-                // It is generally cheaper to change a local and then write back to ref at the end
-                // rather than updating the ref on each operation.
-                char* temp = p;
-
-                StringBuilder sb = StringBuilderCache.Acquire();
-
                 // Skip leading whitespace
-                while (*temp == ' ')
+                while (currentIndex < data.Length && data[currentIndex] == ' ')
                 {
-                    temp++;
+                    currentIndex++;
                 }
 
                 // Start parsing key
-                while (*temp != '=')
+                int start = currentIndex;
+
+                // Parse till '=' is encountered marking end of key.
+                // Key cannot contain whitespace, break if whitespace is found.
+                while (currentIndex < data.Length && data[currentIndex] != '=' && data[currentIndex] != ' ')
                 {
-                    // Key cannot have whitespace
-                    if (*temp == ' ')
-                        break;
+                    currentIndex++;
+                }
 
-                    sb.Append(*temp);
-                    temp++;
+                if (currentIndex == data.Length)
+                {
+                    // Key didn't terminate with '='
+                    parsedIndex = currentIndex;
+                    return null;
                 }
 
-                // Skip trailing whitespace and '='
-                while (*temp == ' ' || *temp == '=')
+                // Record end of key.
+                int length = currentIndex - start;
+                if (data[currentIndex] == ' ')
                 {
-                    temp++;
+                    // Key parsing terminated due to whitespace.
+                    // Parse till '=' is found.
+                    while (currentIndex < data.Length && data[currentIndex] == ' ')
+                    {
+                        currentIndex++;
+                    }
+
+                    if (currentIndex == data.Length || data[currentIndex] != '=')
+                    {
+                        // Key is invalid.
+                        parsedIndex = currentIndex;
+                        return null;
+                    }
                 }
 
-                // Set the ref p to temp
-                p = temp;
+                // Skip trailing whitespace and '='
+                while (currentIndex < data.Length && (data[currentIndex] == ' ' || data[currentIndex] == '='))
```

Comment:
So "key====value" is valid?  This will skip not only trailing whitespace but also multiple equal signs.

Reply:
Yeah, key=value is the valid string, any number of ' ' or '=' is not counted as a valid char in key or unquoted value, the rfc syntax says ```auth-param = token BWS "=" BWS```. Extra '=' is bad syntax, so skipped here.

Reply:
BWS = optional whitespace, which is kept only for compatibility purposes, they have explained it better here https://tools.ietf.org/search/rfc7230#section-3.2.3